collector/thermal_darwin.go

```diff
@@ -0,0 +1,183 @@
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build !notherm
+
+package collector
+
+/*
+#cgo LDFLAGS: -framework IOKit -framework CoreFoundation
+#include <stdio.h>
+#include <CoreFoundation/CoreFoundation.h>
+#include <IOKit/IOKitLib.h>
+#include <IOKit/pwr_mgt/IOPMLib.h>
+#include <IOKit/pwr_mgt/IOPM.h>
+
+struct ref_with_ret {
+    CFDictionaryRef ref;
+    IOReturn ret;
+};
+
+struct ref_with_ret FetchThermal();
+
+struct ref_with_ret FetchThermal() {
+    CFDictionaryRef ref;
+    IOReturn ret;
+    ret = IOPMCopyCPUPowerStatus(&ref);
+    struct ref_with_ret result = {
+            ref,
+            ret,
+    };
+    return result;
+}
+*/
+import "C"
+
+import (
+	"errors"
+	"fmt"
+	"github.com/go-kit/kit/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"unsafe"
+)
+
+type thermCollector struct {
+	cpuSchedulerLimit typedDesc
+	cpuAvailableCPU   typedDesc
+	cpuSpeedLimit     typedDesc
+	logger            log.Logger
+}
+
+const thermal = "thermal"
+
+func init() {
+	registerCollector(thermal, defaultEnabled, NewThermCollector)
+}
+
+// NewThermCollector returns a new Collector exposing current CPU power levels.
+func NewThermCollector(logger log.Logger) (Collector, error) {
+	return &thermCollector{
+		cpuSchedulerLimit: typedDesc{
+			desc: prometheus.NewDesc(
+				prometheus.BuildFQName(namespace, thermal, "cpu_scheduler_limit_ratio"),
+				"Represents the percentage (0-100) of CPU time available. 100% at normal operation. The OS may limit this time for a percentage less than 100%.",
+				nil,
+				nil),
+			valueType: prometheus.GaugeValue,
+		},
+		cpuAvailableCPU: typedDesc{
+			desc: prometheus.NewDesc(
+				prometheus.BuildFQName(namespace, thermal, "cpu_available_cpu"),
+				"Reflects how many, if any, CPUs have been taken offline. Represented as an integer number of CPUs (0 - Max CPUs).",
+				nil,
+				nil,
+			),
+			valueType: prometheus.GaugeValue,
+		},
+		cpuSpeedLimit: typedDesc{
+			desc: prometheus.NewDesc(
+				prometheus.BuildFQName(namespace, thermal, "cpu_speed_limit_ratio"),
+				"Defines the speed & voltage limits placed on the CPU. Represented as a percentage (0-100) of maximum CPU speed.",
+				nil,
+				nil,
+			),
+			valueType: prometheus.GaugeValue,
+		},
+		logger: logger,
+	}, nil
+}
+
+func (c *thermCollector) Update(ch chan<- prometheus.Metric) error {
+	cpuPowerStatus, err := fetchCPUPowerStatus()
+	if err != nil {
+		return err
+	}
+	if value, ok := cpuPowerStatus[(string(C.kIOPMCPUPowerLimitSchedulerTimeKey))]; ok {
+		ch <- c.cpuSchedulerLimit.mustNewConstMetric(float64(value / 100.0))
```

Comment:
You should do `float64(value) / 100.0` to avoid int rounding errors.

Reply:
Oh, I just think it will make implicit type conversion and automatically. Does this will make difference? Please tell me more. 🥺

Reply:
https://play.golang.org/p/C1gfZdWSF2_r